File: 3105.longest-strictly-increasing-or-strictly-decreasing-subarray.cpp
/*
 * @lc app=leetcode id=3105 lang=cpp
 *
 * [3105] Longest Strictly Increasing or Strictly Decreasing Subarray
 */

// @lc code=start
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    /*
    Keep track of last num, result so far, current longest strictly increasing and strictly decreasing subarray ending at num
    */
    int longestMonotonicSubarray(vector<int>& nums) {
        int currentIncr = 1, currentDecr = 1, result = 1, n = nums.size();
        for (int i = 1; i < n; i++) {
            if (nums[i] > nums[i - 1]) {
                currentIncr++;
            } else {
                currentIncr = 1;
            }
            if (nums[i] < nums[i - 1]) {
                currentDecr++;
            } else {
                currentDecr = 1;
            }
            result = max(result, currentIncr);
            result = max(result, currentDecr);
        }
        return result;
    }
};

// @lc code=end

File: 3105.longest-strictly-increasing-or-strictly-decreasing-subarray.py
#
# @lc app=leetcode id=3105 lang=python3
#
# [3105] Longest Strictly Increasing or Strictly Decreasing Subarray
#

# @lc code=start
from typing import List


class Solution:
    '''
    Keep track of last num, result so far, current longest strictly increasing and decreasing subarray ending at num
    '''
    def longestMonotonicSubarray(self, nums: List[int]) -> int:
        currentIncr, currentDecr, result, n = 1, 1, 1, len(nums)
        for i in range(1, n, 1):
            if nums[i] > nums[i - 1]: currentIncr += 1 
            else: currentIncr = 1
            if nums[i] < nums[i - 1]: currentDecr += 1 
            else: currentDecr = 1
            result = max(result, currentIncr)
            result = max(result, currentDecr)
        return result             
        
# @lc code=end

